### PR TITLE
Adding tests for cases when Kafka offsets are NLA

### PR DESCRIPTION
**Pull request by danhermann**

### CODE REVIEW

File: kaldb/src/test/java/com/slack/kaldb/recovery/RecoveryServiceTest.java [hunk 1]
```diff
@@ -164,6 +178,147 @@ public void testShouldHandleRecoveryTask() throws Exception {
...
+        new RecoveryTaskMetadata(
+            topicPartition.topic(),
+            Integer.toString(topicPartition.partition()),
+            startOffset,
+            endOffset,
+            Instant.now().toEpochMilli());
+    assertThat(recoveryService.handleRecoveryTask(recoveryTask)).isTrue();
+    assertThat(getCount(RECORDS_NO_LONGER_AVAILABLE, components.meterRegistry))
+        .isEqualTo(endOffset - startOffset + 1);
+    assertThat(getCount(MESSAGES_RECEIVED_COUNTER, components.meterRegistry)).isEqualTo(0);
```
Comment (mansu) [inline kaldb/src/test/java/com/slack/kaldb/recovery/RecoveryServiceTest.java:249]: Can you please add the asserts from other tests on failed messages being zero and roll over metrics also?

File: kaldb/src/test/java/com/slack/kaldb/recovery/RecoveryServiceTest.java [hunk 2]
```diff
@@ -164,6 +178,147 @@ public void testShouldHandleRecoveryTask() throws Exception {
...
+    RecoveryTaskMetadata recoveryTask =
+        new RecoveryTaskMetadata(
+            topicPartition.topic(),
+            Integer.toString(topicPartition.partition()),
+            startOffset,
+            endOffset,
+            Instant.now().toEpochMilli());
+    assertThat(recoveryService.handleRecoveryTask(recoveryTask)).isTrue();
+    assertThat(getCount(RECORDS_NO_LONGER_AVAILABLE, components.meterRegistry)).isEqualTo(50);
+    assertThat(getCount(MESSAGES_RECEIVED_COUNTER, components.meterRegistry)).isEqualTo(51);
```
Comment (mansu) [inline kaldb/src/test/java/com/slack/kaldb/recovery/RecoveryServiceTest.java:319]: Can you please add the asserts from other tests on failed messages being zero and roll over metrics also?

**Review by mansu — CHANGES_REQUESTED** [review]

File: kaldb/src/main/java/com/slack/kaldb/recovery/RecoveryService.java [hunk 3]
```diff
@@ -245,6 +249,13 @@ boolean handleRecoveryTask(RecoveryTaskMetadata recoveryTaskMetadata) {
               partitionOffsets.endOffset,
               recoveryTaskMetadata.createdTimeEpochMs);
 
+      if (partitionOffsets.startOffset != recoveryTaskMetadata.startOffset
+          || recoveryTaskMetadata.endOffset != partitionOffsets.endOffset) {
+        recoveryRecordsNoLongerAvailable.increment(
```
[the remark this answers is not in the document]
Comment (danhermann) [reply kaldb/src/main/java/com/slack/kaldb/recovery/RecoveryService.java:254]: That happens in `RecoveryService::validateKafkaOffsets`. Would you prefer those logging statements be moved?

File: kaldb/src/test/java/com/slack/kaldb/recovery/RecoveryServiceTest.java (hunk 2 above)
Comment (danhermann) [reply kaldb/src/test/java/com/slack/kaldb/recovery/RecoveryServiceTest.java:330]: https://github.com/slackhq/kaldb/pull/440/commits/76042b28b430df570f2641d26c9f08c0c833acc1

File: kaldb/src/main/java/com/slack/kaldb/recovery/RecoveryService.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mansu) [reply kaldb/src/main/java/com/slack/kaldb/recovery/RecoveryService.java:254]: In that case, we are good.

File: kaldb/src/test/java/com/slack/kaldb/recovery/RecoveryServiceTest.java [hunk 4]
```diff
@@ -164,6 +178,168 @@ public void testShouldHandleRecoveryTask() throws Exception {
...
+    List<SnapshotMetadata> snapshots = snapshotMetadataStore.listSync();
+    assertThat(snapshots.size()).isEqualTo(1);
+    assertThat(blobFs.listFiles(BlobFsUtils.createURI(TEST_S3_BUCKET, "/", ""), true)).isNotEmpty();
+    assertThat(blobFs.exists(URI.create(snapshots.get(0).snapshotPath))).isTrue();
+    assertThat(blobFs.listFiles(URI.create(snapshots.get(0).snapshotPath), false).length)
+        .isGreaterThan(1);
+    assertThat(getCount(MESSAGES_FAILED_COUNTER, meterRegistry)).isEqualTo(0);
+    assertThat(getCount(ROLLOVERS_INITIATED, meterRegistry)).isEqualTo(0);
```
Comment (mansu) [inline kaldb/src/test/java/com/slack/kaldb/recovery/RecoveryServiceTest.java:338]: Something is off here. It seems we have snapshots without roll overs. Is the chunk with 51 messages rolled over?

Comment (danhermann) [reply kaldb/src/test/java/com/slack/kaldb/recovery/RecoveryServiceTest.java:338]: No, the chunk manager is configured to roll at either 10GB or 10k messages.

**Review by mansu — APPROVED** [review]
